File: src/snapshot.ts
// In-memory history snapshot + stable id mapping.
//
// The Burp upstream gives us proxy history as an ordered list with no
// per-entry stable identifier. We assign session-local ids 0..N-1 in the
// order Burp returned them. New entries Burp captured after our last refresh
// keep their relative order and get appended ids — existing ids stay stable,
// which is exactly the contract `view_request(id=42)` relies on.
//
// Refresh policy is incremental:
//   1. Re-fetch starting at offset = N - 1 (last known index).
//   2. The first entry returned must equal our existing entry N-1 — anchor.
//      Everything after is new; append.
//   3. If anchor mismatches (history was cleared / replaced inside Burp),
//      fall back to a full rebuild.

import {
  type ParsedRequest,
  type ParsedResponse,
  parseRequest,
  parseResponse,
  findHeader,
  requestHost,
  requestPath,
  responseContentType,
  responseContentLength,
} from "./format/http-parse.js";

export interface RawEntry {
  request: string;
  response: string;
  notes: string | null;
}

export interface HistoryEntry {
  readonly id: number;
  readonly rawRequest: string;
  readonly rawResponse: string;
  readonly notes: string | null;
  readonly observedAt: number; // ms since epoch
  readonly request: ParsedRequest & { readonly host: string; readonly path: string };
  readonly response: ParsedResponse & {
    readonly contentType: string;
    readonly contentLength: number;
  };
}

function buildEntry(
  id: number,
  raw: RawEntry,
  observedAt: number,
): HistoryEntry {
  const req = parseRequest(raw.request || "");
  const resp = parseResponse(raw.response || "");
  return {
    id,
    rawRequest: raw.request || "",
    rawResponse: raw.response || "",
    notes: raw.notes ?? null,
    observedAt,
    request: { ...req, host: requestHost(req), path: requestPath(req) },
    response: {
      ...resp,
      contentType: responseContentType(resp),
      contentLength: responseContentLength(resp),
    },
  };
}

export function entryTimestamp(e: HistoryEntry): Date {
  // Prefer response Date header (per-entry, server clock); fall back to
  // observedAt (when our middleware ingested the entry).
  const dateHdr = findHeader(e.response.headers, "Date");
  if (dateHdr) {
    const t = Date.parse(dateHdr);
    if (!Number.isNaN(t)) return new Date(t);
  }
  return new Date(e.observedAt);
}

export interface Snapshot {
  readonly entries: readonly HistoryEntry[];
  readonly fetchedAt: number; // monotonic-ish; we use Date.now() — we only do deltas with it
  byId(id: number): HistoryEntry | undefined;
  readonly count: number;
}

function makeSnapshot(entries: readonly HistoryEntry[], fetchedAt: number): Snapshot {
  return {
    entries,
    fetchedAt,
    byId(id) {
      if (id >= 0 && id < entries.length) return entries[id];
      return undefined;
    },
    get count() {
      return entries.length;
    },
  };
}

export type Fetcher = (offset?: number) => Promise<RawEntry[]>;

export class SnapshotStore {
  private readonly fetcher: Fetcher;
  private readonly ttlMs: number;
  private snapshot: Snapshot | null = null;

  constructor(fetcher: Fetcher, ttlSeconds = 30) {
    this.fetcher = fetcher;
    this.ttlMs = ttlSeconds * 1000;
  }

  async get(refresh = false): Promise<Snapshot> {
    const now = Date.now();
    if (
      refresh ||
      this.snapshot === null ||
      now - this.snapshot.fetchedAt > this.ttlMs
    ) {
      await this.refresh();
    }
    return this.snapshot!;
  }

  private async refresh(): Promise<void> {
    const existing = this.snapshot?.entries ?? [];
    if (existing.length === 0) {
      await this.fullRebuild();
      return;
    }

    const anchorIdx = existing.length - 1;
    const page = await this.fetcher(anchorIdx);
    if (page.length === 0) {
      await this.fullRebuild();
      return;
    }

    const anchor = page[0]!;
    const last = existing[anchorIdx]!;
    if (anchor.request !== last.rawRequest || anchor.response !== last.rawResponse) {
      // History reordered or replaced; can't safely append.
      await this.fullRebuild();
      return;
    }

    const newRaws = page.slice(1);
    if (newRaws.length === 0) {
      this.snapshot = makeSnapshot(existing, Date.now());
      return;
    }
    const startId = existing.length;
    const observedAt = Date.now();
    const appended = newRaws.map((r, i) => buildEntry(startId + i, r, observedAt));
    this.snapshot = makeSnapshot([...existing, ...appended], observedAt);
  }

  private async fullRebuild(): Promise<void> {
    const raws = await this.fetcher(0);
    const observedAt = Date.now();
    const entries = raws.map((r, i) => buildEntry(i, r, observedAt));
    this.snapshot = makeSnapshot(entries, observedAt);
  }

  peek(): Snapshot | null {
    return this.snapshot;
  }
}

File: src/burp_mcp_lite/snapshot.py
"""In-memory history snapshot + stable id mapping.

The Burp upstream gives us proxy history as an ordered list with no per-entry
stable identifier. We assign session-local ids `0..N-1` in the order Burp
returned them. New entries Burp captured after our last refresh keep their
relative order and get appended ids — *existing ids stay stable*, which is
exactly the contract `view_request(id=42)` relies on.

Refresh policy is incremental. Burp's proxy history is monotonic-append in
practice, so:

  1. We re-fetch starting at `offset = N - 1` (our last known index).
  2. The first entry returned must equal our existing entry N-1 — that's our
     integrity anchor. Everything after it is new; we append.
  3. If the anchor mismatches (history was cleared or replaced inside Burp),
     we fall back to a full rebuild.

That's one upstream round-trip per refresh, returning ~1 entry when there's
no new traffic and ~K+1 entries when K were captured since last refresh.

Cache invalidation:
  - explicit `refresh=True` runs the incremental refresh (skips TTL check)
  - if the cache is older than `ttl_seconds`, the next access triggers it
  - a cache miss on a specific id does NOT auto-refresh — id stability matters
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Awaitable, Callable, Optional

from .format.http_parse import ParsedRequest, ParsedResponse, parse_request, parse_response


@dataclass
class HistoryEntry:
    """One proxy history entry, parsed and ready for projection.

    `raw_request` / `raw_response` are kept verbatim so we can re-slice without
    losing fidelity. The parsed views are derived once.
    """

    id: int
    raw_request: str
    raw_response: str
    notes: Optional[str] = None
    request: ParsedRequest = field(init=False)
    response: ParsedResponse = field(init=False)

    def __post_init__(self) -> None:
        self.request = parse_request(self.raw_request or "")
        self.response = parse_response(self.raw_response or "")


@dataclass
class Snapshot:
    entries: list[HistoryEntry]
    fetched_at: float

    def by_id(self, id_: int) -> Optional[HistoryEntry]:
        if 0 <= id_ < len(self.entries):
            return self.entries[id_]
        return None

    @property
    def count(self) -> int:
        return len(self.entries)


# Type for the upstream fetcher: `fetch(offset=N)` returns the slice of history
# starting at `offset`. The upstream is responsible for any internal paging and
# for capping at its own max_entries.
RawEntry = tuple[str, str, Optional[str]]
Fetcher = Callable[..., Awaitable[list[RawEntry]]]


class SnapshotStore:
    """Owns the current snapshot and the refresh policy.

    Single-process, single-coroutine ownership assumed (we're stdio-driven,
    requests serialize). No locking.
    """

    def __init__(self, fetcher: Fetcher, ttl_seconds: float = 30.0) -> None:
        self._fetch = fetcher
        self._ttl = ttl_seconds
        self._snapshot: Optional[Snapshot] = None

    async def get(self, refresh: bool = False) -> Snapshot:
        now = time.monotonic()
        if (
            refresh
            or self._snapshot is None
            or (now - self._snapshot.fetched_at) > self._ttl
        ):
            await self._refresh()
        assert self._snapshot is not None
        return self._snapshot

    async def _refresh(self) -> None:
        existing = self._snapshot.entries if self._snapshot else []
        if not existing:
            await self._full_rebuild()
            return

        # Anchor probe + tail in one call. Page[0] should equal entry N-1.
        anchor_idx = len(existing) - 1
        page = await self._fetch(offset=anchor_idx)
        if not page:
            # Burp now has fewer entries than us — history was cleared.
            await self._full_rebuild()
            return

        anchor_req, anchor_resp, _ = page[0]
        last = existing[anchor_idx]
        if anchor_req != last.raw_request or anchor_resp != last.raw_response:
            # History was reordered or replaced; can't safely append.
            await self._full_rebuild()
            return

        new_raws = page[1:]
        if not new_raws:
            # No new traffic; just bump the cache timestamp.
            self._snapshot = Snapshot(entries=existing, fetched_at=time.monotonic())
            return

        start_id = len(existing)
        appended = [
            HistoryEntry(id=start_id + i, raw_request=r, raw_response=s, notes=n)
            for i, (r, s, n) in enumerate(new_raws)
        ]
        self._snapshot = Snapshot(
            entries=existing + appended, fetched_at=time.monotonic()
        )

    async def _full_rebuild(self) -> None:
        raws = await self._fetch(offset=0)
        entries = [
            HistoryEntry(id=i, raw_request=req, raw_response=resp, notes=notes)
            for i, (req, resp, notes) in enumerate(raws)
        ]
        self._snapshot = Snapshot(entries=entries, fetched_at=time.monotonic())

    def peek(self) -> Optional[Snapshot]:
        return self._snapshot
